keps/sig-api-machinery/3285-subresource-finalizers/README.md

```diff
@@ -0,0 +1,812 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a way to set different RBACs to `finalizers` field and its entire resource.
+
+## Motivation
+
+`Finalizers` is used by controllers as a pre-deletion hook.
+There is no subresource for `finalizers`, so the only way to grant an actor access to them is to grant access to the whole object.
+On the other hand, some controllers only need to update `finalizers` field and some of its subresources, and don't need to update its entire resource.
+To implement least privilege for such controllers, there should be a way to only allow updating `finalizers`.
+
+### Goals
+
+- Provide a way to set a different RBAC to `finalizers` from its entire resource
+- Provide a way to audit and update only `metadata.finalizers` field
+
+### Non-Goals
+
+- Provide a way to set a different RBAC to other than `finalizers` from its entire resource
+- Provide a way to allow updating its entire resource  but disallow updating `finalizers` field
+
+## Proposal
+
+A new REST endpoint `<resource>/finalizers` is added automatically to all resources as an alternative way to access the `metadata.finalizers` fielld, which allows separate RBAC.
+The existing method to acceess `metadata.finalizers` field (through the main `<resource>` endpoint) remains unchanged.
+
+### User Stories (Optional)
+
+#### Story 1
+
+There is a controller that only updates `status` field and `finalizers` field of a CRD. Administrators of the controller can allow the controller to update only these particular fields.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Name collision of the endpoints is needed to be considered.
+- Other existing fields may require the same alternative way to access.
+For example, [candidates](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go) in metadata fields are:
+  - `labels`
+  - `annotations`
+  - `ownerReferences`
+  - `status/conditions`
```

Comment:
This is too many separate things / permissions if we do it for every metadata field. If we really need to do this, I think it'd be more efficient and a better user experience to let us RBAC individual fields somehow.

Reply:
I could see `/metadata` as a single subresource, but it feels unsatisfying.

Spelling out exactly which parts of a resource a controller needs seems like a decent goal, doesn't it?

Reply:
Making permissions somehow a la carte seems like the important goal. Making a subresource per permission seems like it will make many more than a reasonable amount of subresources.

Reply:
@deads2k may disagree with me here.

Reply:
Per-field permissions would be really handy in some cases. Especially if they had awareness of K8s idioms like lists of "keyed" resources that are actually maps. Then, for example, we could grant nodes permission to update only specific taints, rather than today's heavy and opaque workaround of granting update on node /status so the node can create a node condition, and then building a controller to reconcile that condition to a specific taint. 

Does the strategic merge patch infrastructure open up opportunities/easy ways to implement per-field permissions that didn't previously exist? (vaguely, if we're more diff-aware now, maybe it's easier to implement RBAC rules that describe which diffs are allowed)

Reply:
Also maybe per-field permissions are helpful to users of SSA since you can make the field owners explicit in RBAC.

Reply:
> Does the server side apply infrastructure open up opportunities/easy ways to implement per-field permissions that didn't previously exist? 

We offer a fieldpath concept that is aware of k8s idoms like named list items.

I don't think SSA conflicts are quite the right thing for this, but we could definitely reuse building blocks.

Reply:
@mikedanese @liggitt @ritazh  Do you think there'd be any interest from SIG-Auth folks in working on granular RBAC?